File: static/js/enterprise-3d-rooms.js
// Enterprise 3D Rooms Management
class Enterprise3DRooms {
    constructor() {
        this.currentEnterprise = null;
        this.availableRooms = [];
        this.defaultRooms = [
            {
                id: 'bathroom',
                name: 'Bathroom',
                thumbnail: '/static/3drooms/Bathroom/bathroom.jpg',
                model_path: '/static/3drooms/Bathroom/bathroom.gltf',
                is_default: true
            },
            {
                id: 'kitchen',
                name: 'Kitchen', 
                thumbnail: '/static/3drooms/Kitchen/kitchen.jpg',
                model_path: '/static/3drooms/Kitchen/kitchen.gltf',
                is_default: true
            },
            {
                id: 'living-room',
                name: 'Living Room',
                thumbnail: '/static/3drooms/Living Room/living-room.jpg',
                model_path: '/static/3drooms/Living Room/living-room.gltf',
                is_default: true
            },
            {
                id: 'bedroom',
                name: 'Bedroom',
                thumbnail: '/static/3drooms/Bedroom/bedroom.jpg',
                model_path: '/static/3drooms/Bedroom/bedroom.gltf',
                is_default: true
            }
        ];
        this.init();
    }

    init() {
        console.log('Initializing Enterprise3DRooms...');
        this.detectEnterprise();
        
        // Wait for DOM to be fully ready, then load rooms
        if (document.readyState === 'loading') {
            document.addEventListener('DOMContentLoaded', () => {
                this.loadRoomsWhenReady();
            });
        } else {
            this.loadRoomsWhenReady();
        }
        
        this.setupEventListeners();
    }
    
    loadRoomsWhenReady() {
        if (this.currentEnterprise) {
            console.log('Enterprise mode detected, loading enterprise rooms');
            this.loadAvailableRooms();
        } else {
            console.log('Main site mode, keeping default rooms');
            this.isInitialized = true;
        }
    }

    detectEnterprise() {
        // Check if we're on an enterprise subdomain
        const hostname = window.location.hostname;
        console.log('Current hostname:', hostname);
        
        // Check for subdomain pattern
        const parts = hostname.split('.');
        console.log('Hostname parts:', parts);
        
        if (parts.length >= 2) {
            // Check if first part is not 'www' and not 'localhost'
            if (parts[0] !== 'www' && parts[0] !== 'localhost' && hostname !== 'localhost') {
                this.currentEnterprise = parts[0];
                console.log('Enterprise detected:', this.currentEnterprise);
            } else if (hostname === 'localhost' || hostname.includes('localhost')) {
                // For localhost testing, check if there's a subdomain in the host header
                // This would be set by nginx proxy
                this.currentEnterprise = null;
                console.log('Localhost detected - no enterprise');
            }
        }
        
        // Also check body data attributes as fallback
        const userEnterprise = document.body.getAttribute('data-user-enterprise');
        if (userEnterprise && userEnterprise.trim()) {
            this.currentEnterprise = userEnterprise.trim();
            console.log('Enterprise from body attribute:', this.currentEnterprise);
        }
    }

    async loadAvailableRooms() {
        console.log('Loading available rooms for enterprise:', this.currentEnterprise);
        
        try {
            let apiUrl = '/api/enterprise/3d-rooms';
            
            // Add enterprise parameter if we have one
            if (this.currentEnterprise) {
                apiUrl += `?enterprise=${this.currentEnterprise}`;
            }
            
            console.log('Fetching from:', apiUrl);
            
            const response = await fetch(apiUrl, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    // Add host header for enterprise detection
                    ...(this.currentEnterprise && {
                        'X-Enterprise': this.currentEnterprise
                    })
                }
            });
            
            console.log('API response status:', response.status);
            
            if (response.ok) {
                const data = await response.json();
                console.log('API response data:', data);
                
                if (data.rooms && data.rooms.length > 0) {
                    this.availableRooms = data.rooms;
                    console.log('Loaded rooms:', this.availableRooms);
                } else {
                    console.log('No rooms in response, using defaults');
                    this.availableRooms = this.defaultRooms;
                }
            } else {
                console.warn('API request failed with status:', response.status);
                const errorText = await response.text();
                console.warn('Error response:', errorText);
                this.availableRooms = this.defaultRooms;
            }
        } catch (error) {
            console.error('Error loading 3D rooms:', error);
            this.availableRooms = this.defaultRooms;
        }
        
        // Always render the rooms after loading
        this.renderRoomTemplates();
    }

    renderRoomTemplates() {
        console.log('Rendering room templates with rooms:', this.availableRooms);
        
        // Find the room templates container
        const container = document.querySelector('#room-templates-container');
        if (!container) {
            console.error('Room templates container not found');
            return;
        }

        // Clear existing content for enterprise mode
        if (this.currentEnterprise) {
            console.log('Clearing existing room templates for enterprise');
            container.innerHTML = '';
        }

        // Render each room
        this.availableRooms.forEach(room => {
            const roomCard = this.createRoomCard(room);
            container.appendChild(roomCard);
        });

        // Add upload button for enterprise admins
        if (this.currentEnterprise && this.isEnterpriseAdmin()) {
            const uploadCard = this.createUploadCard();
            container.appendChild(uploadCard);
        }
        
        console.log('Room templates rendered successfully');
    }


    createRoomCard(room) {
        const colDiv = document.createElement('div');
        colDiv.className = 'col-md-4 mb-4';

        // Use the room's model_path for loading, fallback to id for default rooms
        const modelPath = room.model_path || `/static/3drooms/${room.name}/${room.id}.gltf`;
        
        colDiv.innerHTML = `
            <div class="card room-template-card" onclick="enterprise3DRooms.load3DTemplate('${room.id}', '${modelPath}')">
                <div class="card-img-container">
                    <img src="${room.thumbnail_path || room.thumbnail}" class="card-img-top" alt="${room.name}" 
                        onerror="this.src='/static/img/placeholder-room.jpg'">
                    ${!room.is_default ? '<span class="badge bg-success position-absolute top-0 end-0 m-2">Custom</span>' : ''}
                </div>
                <div class="card-body text-center">
                    <h6 class="card-title mb-0">${room.name}</h6>
                    ${!room.is_default && this.isEnterpriseAdmin() ? 
                        `<button class="btn btn-sm btn-outline-danger mt-2" onclick="event.stopPropagation(); enterprise3DRooms.deleteRoom('${room.id}')">
                            <i class="fas fa-trash"></i> Delete
                        </button>` : ''
                    }
                </div>
            </div>
        `;

        return colDiv;
    }


    createUploadCard() {
        const colDiv = document.createElement('div');
        colDiv.className = 'col-md-4 mb-4';

        colDiv.innerHTML = `
            <div class="card room-template-card upload-card" onclick="enterprise3DRooms.showUploadModal()">
                <div class="card-img-container d-flex align-items-center justify-content-center bg-light">
                    <i class="fas fa-plus fa-3x text-muted"></i>
                </div>
                <div class="card-body text-center">
                    <h6 class="card-title mb-0">Upload New Room</h6>
                    <small class="text-muted">Add custom 3D room</small>
                </div>
            </div>
        `;

        return colDiv;
    }

    load3DTemplate(roomId, modelPath) {
        console.log('Loading 3D template:', roomId, 'with model path:', modelPath);
        
        // Call the original 3D visualizer function
        if (typeof window.load3DTemplate === 'function') {
            // If modelPath is provided, use it; otherwise use roomId for default rooms
            if (modelPath) {
                // For custom rooms, we need to modify the load3DTemplate function to accept model path
                window.load3DTemplate(roomId, modelPath);
            } else {
                window.load3DTemplate(roomId);
            }
        } else if (typeof load3DTemplate === 'function') {
            if (modelPath) {
                load3DTemplate(roomId, modelPath);
            } else {
                load3DTemplate(roomId);
            }
        } else {
            console.error('load3DTemplate function not found');
        }
    }

    isEnterpriseAdmin() {
        // Check if current user is admin of this enterprise
        const userRole = document.body.getAttribute('data-user-role');
        const userEnterprise = document.body.getAttribute('data-user-enterprise');
        
        console.log('Checking admin status:', { userRole, userEnterprise, currentEnterprise: this.currentEnterprise });
        
        return (userRole === 'enterprise_admin' || userRole === 'superadmin') && 
               (userEnterprise === this.currentEnterprise || userRole === 'superadmin');
    }

    showUploadModal() {
        // Create and show upload modal
        const modal = document.getElementById('upload3DRoomModal');
        if (modal) {
            const bootstrapModal = new bootstrap.Modal(modal);
            bootstrapModal.show();
        } else {
            this.createUploadModal();
        }
    }

    createUploadModal() {
        const modalHTML = `
            <div class="modal fade" id="upload3DRoomModal" tabindex="-1">
                <div class="modal-dialog modal-lg">
                    <div class="modal-content">
                        <div class="modal-header">
                            <h5 class="modal-title">Upload 3D Room</h5>
                            <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                        </div>
                        <div class="modal-body">
                            <form id="upload3DRoomForm" enctype="multipart/form-data">
                                <div class="mb-3">
                                    <label for="roomName" class="form-label">Room Name</label>
                                    <input type="text" class="form-control" id="roomName" name="name" required>
                                </div>
                                
                                <div class="mb-3">
                                    <label for="roomThumbnail" class="form-label">Thumbnail Image</label>
                                    <input type="file" class="form-control" id="roomThumbnail" name="thumbnail" accept="image/*" required>
                                    <small class="form-text text-muted">Upload a preview image for the room</small>
                                </div>
                                
                                <div class="mb-3">
                                    <label for="room3DModel" class="form-label">3D Model File</label>
                                    <input type="file" class="form-control" id="room3DModel" name="model" accept=".gltf,.glb" required>
                                    <small class="form-text text-muted">Supported formats: .gltf, .glb</small>
                                </div>
                                
                                <div class="mb-3">
                                    <label for="roomDescription" class="form-label">Description (Optional)</label>
                                    <textarea class="form-control" id="roomDescription" name="description" rows="3"></textarea>
                                </div>
                            </form>
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Cancel</button>
                            <button type="button" class="btn btn-primary" onclick="enterprise3DRooms.uploadRoom()">
                                <i class="fas fa-upload"></i> Upload Room
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        `;

        document.body.insertAdjacentHTML('beforeend', modalHTML);
        const modal = new bootstrap.Modal(document.getElementById('upload3DRoomModal'));
        modal.show();
    }

    async uploadRoom() {
        const form = document.getElementById('upload3DRoomForm');
        const formData = new FormData(form);
        
        // Add enterprise info
        formData.append('enterprise', this.currentEnterprise);

        try {
            const response = await fetch('/api/enterprise/3d-rooms/upload', {
                method: 'POST',
                body: formData
            });

            if (response.ok) {
                const result = await response.json();
                console.log('Room uploaded successfully:', result);
                
                // Close modal
                const modal = bootstrap.Modal.getInstance(document.getElementById('upload3DRoomModal'));
                modal.hide();
                
                // Reload rooms
                await this.loadAvailableRooms();
                
                // Show success message
                this.showMessage('3D Room uploaded successfully!', 'success');
            } else {
                const error = await response.json();
                throw new Error(error.message || 'Upload failed');
            }
        } catch (error) {
            console.error('Error uploading room:', error);
            this.showMessage('Error uploading room: ' + error.message, 'error');
        }
    }

    async deleteRoom(roomId) {
        if (!confirm('Are you sure you want to delete this 3D room?')) {
            return;
        }

        try {
            const response = await fetch(`/api/enterprise/3d-rooms/${roomId}`, {
                method: 'DELETE'
            });

            if (response.ok) {
                await this.loadAvailableRooms();
                this.showMessage('3D Room deleted successfully!', 'success');
            } else {
                const error = await response.json();
                throw new Error(error.message || 'Delete failed');
            }
        } catch (error) {
            console.error('Error deleting room:', error);
            this.showMessage('Error deleting room: ' + error.message, 'error');
        }
    }

    showMessage(message, type) {
        // Create toast notification
        const toast = document.createElement('div');
        toast.className = `alert alert-${type === 'error' ? 'danger' : 'success'} position-fixed`;
        toast.style.cssText = 'top: 20px; right: 20px; z-index: 9999; min-width: 300px;';
        toast.innerHTML = `
            ${message}
            <button type="button" class="btn-close float-end" onclick="this.parentElement.remove()"></button>
        `;
        
        document.body.appendChild(toast);
        
        // Auto remove after 5 seconds
        setTimeout(() => {
            if (toast.parentElement) {
                toast.remove();
            }
        }, 5000);
    }

    setupEventListeners() {
        // Listen for tab changes to reload rooms
        const threeDTab = document.getElementById('threed-visualizer-tab');
        if (threeDTab) {
            threeDTab.addEventListener('shown.bs.tab', () => {
                console.log('3D tab shown');
                if (this.currentEnterprise && !this.isInitialized) {
                    console.log('Loading enterprise rooms on tab show');
                    this.loadAvailableRooms();
                }
            });
        }
    }
}

// Initialize when DOM is loaded
document.addEventListener('DOMContentLoaded', function() {
    console.log('DOM loaded, initializing Enterprise3DRooms');
    window.enterprise3DRooms = new Enterprise3DRooms();
});

// Make it globally available
window.Enterprise3DRooms = Enterprise3DRooms;

File: routes/enterprise_3d_rooms.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form, Request
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from database.database import get_db
from database.models import Enterprise3DRoom, Enterprise, EnterpriseUser
from enterprise import get_enterprise, get_enterprise_user
from auth.dependencies import get_current_user
import os
import uuid
from PIL import Image
import shutil

router = APIRouter()

# Default 3D rooms (same as your existing default rooms)
DEFAULT_3D_ROOMS = [
    {
        "id": "bathroom",
        "name": "Bathroom",
        "category": "bathroom",
        "thumbnail": "/static/3drooms/Bathroom/bathroom.jpg",
        "model_path": "/static/3drooms/Bathroom/bathroom.gltf",
        "is_default": True
    },
    {
        "id": "kitchen", 
        "name": "Kitchen",
        "category": "kitchen",
        "thumbnail": "/static/3drooms/Kitchen/kitchen.jpg",
        "model_path": "/static/3drooms/Kitchen/kitchen.gltf",
        "is_default": True
    },
    {
        "id": "living-room",
        "name": "Living Room", 
        "category": "living_room",
        "thumbnail": "/static/3drooms/Living Room/living-room.jpg",
        "model_path": "/static/3drooms/Living Room/living-room.gltf",
        "is_default": True
    },
    {
        "id": "bedroom",
        "name": "Bedroom",
        "category": "bedroom", 
        "thumbnail": "/static/3drooms/Bedroom/bedroom.jpg",
        "model_path": "/static/3drooms/Bedroom/bedroom.gltf",
        "is_default": True
    }
]

@router.get("/api/enterprise/3d-rooms")
async def get_enterprise_3d_rooms(
    request: Request,
    enterprise_param: str = Query(None, alias="enterprise"),
    db: Session = Depends(get_db),
    enterprise = Depends(get_enterprise)
):
    """Get 3D rooms for current enterprise context"""
    try:
        # Try to get enterprise from multiple sources
        current_enterprise = enterprise
        
        # If no enterprise from dependency, try from query parameter
        if not current_enterprise and enterprise_param:
            current_enterprise = db.query(Enterprise).filter(
                Enterprise.subdomain == enterprise_param
            ).first()
        
        # Also check X-Enterprise header
        if not current_enterprise:
            enterprise_header = request.headers.get("X-Enterprise")
            if enterprise_header:
                current_enterprise = db.query(Enterprise).filter(
                    Enterprise.subdomain == enterprise_header
                ).first()
        
        print(f"Enterprise detection result: {current_enterprise.subdomain if current_enterprise else 'None'}")
        
        if current_enterprise:
            # Get enterprise-specific 3D rooms
            enterprise_rooms = db.query(Enterprise3DRoom).filter(
                Enterprise3DRoom.enterprise_id == current_enterprise.id,
                Enterprise3DRoom.is_active == True
            ).all()
            
            print(f"Found {len(enterprise_rooms)} enterprise rooms")
            
            rooms = []
            for room in enterprise_rooms:
                rooms.append({
                    "id": str(room.id),
                    "name": room.name,
                    "category": room.category,
                    "thumbnail": room.thumbnail_path,
                    "model_path": room.model_path,
                    "description": room.description,
                    "is_default": False
                })
            
            # If no enterprise rooms, return defaults
            if not rooms:
                print("No enterprise rooms found, returning defaults")
                rooms = DEFAULT_3D_ROOMS
                
        else:
            # Main site - return default rooms
            print("No enterprise context, returning default rooms")
            rooms = DEFAULT_3D_ROOMS
        
        return JSONResponse(content={"rooms": rooms})
        
    except Exception as e:
        print(f"Error getting 3D rooms: {str(e)}")
        import traceback
        traceback.print_exc()
        return JSONResponse(content={"rooms": DEFAULT_3D_ROOMS})


@router.post("/api/enterprise/3d-rooms/upload")
async def upload_enterprise_3d_room(
    request: Request,
    name: str = Form(...),
    category: str = Form(...),
    description: str = Form(None),
    thumbnail: UploadFile = File(...),
    model: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
    enterprise = Depends(get_enterprise),
    enterprise_user = Depends(get_enterprise_user)
):
    """Upload a new 3D room for enterprise"""
    
    # Check permissions
    if not current_user:
        raise HTTPException(status_code=401, detail="Authentication required")
    
    if not enterprise:
        raise HTTPException(status_code=400, detail="Enterprise context required")
    
    # Check if user is admin of this enterprise
    if not enterprise_user or enterprise_user.role not in ["admin", "owner"]:
        if current_user.role != "superadmin":
            raise HTTPException(status_code=403, detail="Admin access required")
    
    try:
        # Create directories
        enterprise_3d_dir = f"static/enterprise_3d_rooms/{enterprise.id}"
        os.makedirs(enterprise_3d_dir, exist_ok=True)
        
        # Save thumbnail
        thumbnail_ext = os.path.splitext(thumbnail.filename)[1]
        thumbnail_filename = f"{uuid.uuid4().hex}{thumbnail_ext}"
        thumbnail_path = os.path.join(enterprise_3d_dir, thumbnail_filename)
        
        with open(thumbnail_path, "wb") as buffer:
            shutil.copyfileobj(thumbnail.file, buffer)
        
        # Resize thumbnail if needed
        with Image.open(thumbnail_path) as img:
            if img.width > 400 or img.height > 300:
                img.thumbnail((400, 300), Image.Resampling.LANCZOS)
                img.save(thumbnail_path)
        
        # Save 3D model
        model_ext = os.path.splitext(model.filename)[1]
        model_filename = f"{uuid.uuid4().hex}{model_ext}"
        model_path = os.path.join(enterprise_3d_dir, model_filename)
        
        with open(model_path, "wb") as buffer:
            shutil.copyfileobj(model.file, buffer)
        
        # Create database record
        new_room = Enterprise3DRoom(
            enterprise_id=enterprise.id,
            name=name,
            category=category,
            model_path=f"/{model_path}",
            thumbnail_path=f"/{thumbnail_path}",
            description=description,
            is_active=True
        )
        
        db.add(new_room)
        db.commit()
        db.refresh(new_room)
        
        return JSONResponse(content={
            "success": True,
            "message": "3D room uploaded successfully",
            "room_id": new_room.id
        })
        
    except Exception as e:
        db.rollback()
        print(f"Error uploading 3D room: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Upload failed: {str(e)}")

@router.delete("/api/enterprise/3d-rooms/{room_id}")
async def delete_enterprise_3d_room(
    room_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user),
    enterprise = Depends(get_enterprise),
    enterprise_user = Depends(get_enterprise_user)
):
    """Delete a 3D room"""
    
    # Check permissions
    if not current_user:
        raise HTTPException(status_code=401, detail="Authentication required")
    
    if not enterprise:
        raise HTTPException(status_code=400, detail="Enterprise context required")
    
    if not enterprise_user or enterprise_user.role not in ["admin", "owner"]:
        if current_user.role != "superadmin":
            raise HTTPException(status_code=403, detail="Admin access required")
    
    # Get the room
    room = db.query(Enterprise3DRoom).filter(
        Enterprise3DRoom.id == room_id,
        Enterprise3DRoom.enterprise_id == enterprise.id
    ).first()
    
    if not room:
        raise HTTPException(status_code=404, detail="3D room not found")
    
    try:
        # Delete files
        if os.path.exists(room.thumbnail_path.lstrip('/')):
            os.remove(room.thumbnail_path.lstrip('/'))
        
        if os.path.exists(room.model_path.lstrip('/')):
            os.remove(room.model_path.lstrip('/'))
        
        # Delete database record
        db.delete(room)
        db.commit()
        
        return JSONResponse(content={
            "success": True,
            "message": "3D room deleted successfully"
        })
        
    except Exception as e:
        db.rollback()
        print(f"Error deleting 3D room: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Delete failed: {str(e)}")
